lib/psabpf_pipeline.c

```diff
@@ -113,175 +160,179 @@ static int xdp_port_add(__u32 pipeline_id, char *intf)
     sprintf(cmd, "tc qdisc add dev %s clsact", intf);
     system(cmd);
     memset(cmd, 0, sizeof(cmd));
-    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%d/%s",
-            intf, BPF_FS, PIPELINE_PREFIX, pipeline_id, TC_INGRESS_PROG);
+    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%u/%s",
+            intf, BPF_FS, PIPELINE_PREFIX, ctx->pipeline_id, TC_INGRESS_PROG);
     system(cmd);
     memset(cmd, 0, sizeof(cmd));
-    sprintf(cmd, "tc filter add dev %s egress bpf da fd %s/%s%d/%s",
-            intf, BPF_FS, PIPELINE_PREFIX, pipeline_id, TC_EGRESS_PROG);
+    sprintf(cmd, "tc filter add dev %s egress bpf da fd %s/%s%u/%s",
+            intf, BPF_FS, PIPELINE_PREFIX, ctx->pipeline_id, TC_EGRESS_PROG);
     system(cmd);
 
-    return 0;
+    return NO_ERROR;
 }
 
-static int tc_port_add(__u32 pipeline_id, char *intf)
+static int tc_port_add(psabpf_context_t *ctx, const char *intf, int ifindex)
 {
+    int xdp_helper_fd;
+
+    int ret = xdp_attach_prog_to_port(&xdp_helper_fd, ctx, ifindex, XDP_HELPER_PROG);
+    if (ret != NO_ERROR)
+        return ret;
+    close_object_fd(&xdp_helper_fd);
+
     /* FIXME: using bash command only for the PoC purpose */
     char cmd[256];
-    sprintf(cmd, "bpftool net attach xdp pinned %s/%s%d/%s dev %s overwrite",
-            BPF_FS, PIPELINE_PREFIX, pipeline_id, XDP_HELPER_PROG, intf);
-    system(cmd);
     sprintf(cmd, "tc qdisc add dev %s clsact", intf);
     system(cmd);
     memset(cmd, 0, sizeof(cmd));
-    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%d/%s",
-            intf, BPF_FS, PIPELINE_PREFIX, pipeline_id, TC_INGRESS_PROG);
+    sprintf(cmd, "tc filter add dev %s ingress bpf da fd %s/%s%u/%s",
```

Comment:
We can add `TODO` here to use libbpf for installing TC programs, instead of `tc filter`